File: electron/main/ai/anthropic.ts
import {
  AiEvent,
  AiProvider,
  CompleteRequest,
  EventSink,
  JsonValue,
  ModelInfo,
  estimateCost,
} from './provider'

const ANTHROPIC_VERSION = '2023-06-01'
const BASE = 'https://api.anthropic.com/v1'

export class AnthropicProvider implements AiProvider {
  private apiKey: string

  constructor(apiKey: string) {
    this.apiKey = apiKey
  }

  async streamComplete(
    req: CompleteRequest,
    sink: EventSink,
    signal: AbortSignal
  ): Promise<void> {
    const body: { [k: string]: JsonValue } = {
      model: req.model,
      max_tokens: req.maxTokens ?? 4096,
      stream: true,
      messages: req.messages.map((m) => ({ role: m.role, content: m.content })),
    }
    if (req.system != null) body.system = req.system
    if (req.temperature != null) body.temperature = req.temperature

    let resp: Response
    try {
      resp = await fetch(`${BASE}/messages`, {
        method: 'POST',
        headers: {
          'x-api-key': this.apiKey,
          'anthropic-version': ANTHROPIC_VERSION,
          'content-type': 'application/json',
          accept: 'text/event-stream',
        },
        body: JSON.stringify(body),
        signal,
      })
    } catch (e) {
      if (signal.aborted) return
      throw new Error(`anthropic request: ${e instanceof Error ? e.message : String(e)}`)
    }

    if (!resp.ok) {
      const text = await resp.text().catch(() => '')
      throw new Error(`anthropic ${resp.status}: ${text}`)
    }
    if (!resp.body) {
      throw new Error('anthropic: empty response body')
    }

    let inTokens = 0
    let outTokens = 0
    const reader = resp.body.getReader()
    const decoder = new TextDecoder('utf-8')
    let buf = ''
    let donned = false

    const emitDone = () => {
      if (donned) return
      donned = true
      const cost = estimateCost('anthropic', req.model, inTokens, outTokens)
      sink({ kind: 'done', value: { inTokens, outTokens, costUsd: cost } })
    }

    try {
      while (true) {
        if (signal.aborted) {
          try { await reader.cancel() } catch {}
          return
        }
        let chunk: { done: boolean; value?: Uint8Array }
        try {
          chunk = await reader.read()
        } catch (e) {
          if (signal.aborted) return
          throw new Error(`read sse: ${e instanceof Error ? e.message : String(e)}`)
        }
        if (chunk.done) break
        if (!chunk.value) continue
        buf += decoder.decode(chunk.value, { stream: true })

        let nl: number
        while ((nl = buf.indexOf('\n')) >= 0) {
          const rawLine = buf.slice(0, nl)
          buf = buf.slice(nl + 1)
          const line = rawLine.replace(/\r$/, '')
          if (!line.startsWith('data: ')) continue
          const payload = line.slice(6)
          if (payload.length === 0) continue
          let v: { [k: string]: JsonValue }
          try {
            v = JSON.parse(payload) as { [k: string]: JsonValue }
          } catch {
            continue
          }
          const t = typeof v.type === 'string' ? (v.type as string) : ''
          if (t === 'content_block_delta') {
            const delta = v.delta as { [k: string]: JsonValue } | undefined
            const text = delta?.text
            if (typeof text === 'string') {
              const ev: AiEvent = { kind: 'delta', value: text }
              sink(ev)
            }
          } else if (t === 'message_start') {
            const msg = v.message as { [k: string]: JsonValue } | undefined
            const u = msg?.usage as { [k: string]: JsonValue } | undefined
            if (u) {
              if (typeof u.input_tokens === 'number') inTokens = u.input_tokens as number
              if (typeof u.output_tokens === 'number') outTokens = u.output_tokens as number
            }
          } else if (t === 'message_delta') {
            const u = v.usage as { [k: string]: JsonValue } | undefined
            if (u && typeof u.output_tokens === 'number') {
              outTokens = u.output_tokens as number
            }
          } else if (t === 'message_stop') {
            emitDone()
            return
          }
        }
      }
    } finally {
    }

    emitDone()
  }

  async listModels(): Promise<ModelInfo[]> {
    const resp = await fetch(`${BASE}/models`, {
      headers: {
        'x-api-key': this.apiKey,
        'anthropic-version': ANTHROPIC_VERSION,
      },
    })
    if (!resp.ok) {
      const text = await resp.text().catch(() => '')
      throw new Error(`anthropic models ${resp.status}: ${text}`)
    }
    const parsed = (await resp.json()) as {
      data: Array<{ id: string; display_name?: string }>
    }
    return parsed.data.map((m) => ({
      id: m.id,
      name: m.display_name ?? m.id,
    }))
  }
}

File: src-tauri/src/ai/anthropic.rs
use std::io::{BufRead, BufReader};
use std::sync::atomic::{AtomicBool, Ordering};
use std::time::Duration;

use anyhow::{anyhow, Context, Result};
use reqwest::blocking::Client;
use serde::Deserialize;
use serde_json::json;

use super::provider::{
    estimate_cost, AiEvent, AiProvider, CompleteRequest, EventSink, ModelInfo, Usage,
};

const ANTHROPIC_VERSION: &str = "2023-06-01";
const BASE: &str = "https://api.anthropic.com/v1";

pub struct AnthropicProvider {
    api_key: String,
    client: Client,
}

impl AnthropicProvider {
    pub fn new(api_key: String) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(300))
            .build()
            .expect("reqwest client");
        Self { api_key, client }
    }
}

impl AiProvider for AnthropicProvider {
    fn stream_complete(
        &self,
        req: &CompleteRequest,
        sink: EventSink,
        cancel: &AtomicBool,
    ) -> Result<()> {
        let mut body = json!({
            "model": req.model,
            "max_tokens": req.max_tokens.unwrap_or(4096),
            "stream": true,
            "messages": req.messages.iter().map(|m| json!({
                "role": m.role,
                "content": m.content,
            })).collect::<Vec<_>>(),
        });
        if let Some(sys) = &req.system {
            body["system"] = json!(sys);
        }
        if let Some(t) = req.temperature {
            body["temperature"] = json!(t);
        }

        let resp = self
            .client
            .post(format!("{BASE}/messages"))
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", ANTHROPIC_VERSION)
            .header("content-type", "application/json")
            .json(&body)
            .send()
            .context("anthropic request")?;

        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().unwrap_or_default();
            return Err(anyhow!("anthropic {}: {}", status, text));
        }

        let mut reader = BufReader::new(resp);
        let mut line = String::new();
        let mut in_tokens: u32 = 0;
        let mut out_tokens: u32 = 0;

        loop {
            if cancel.load(Ordering::SeqCst) {
                return Ok(());
            }
            line.clear();
            let n = reader.read_line(&mut line).context("read sse line")?;
            if n == 0 {
                break;
            }
            let trimmed = line.trim_end_matches(['\r', '\n']);
            if let Some(payload) = trimmed.strip_prefix("data: ") {
                if payload.is_empty() {
                    continue;
                }
                let v: serde_json::Value = match serde_json::from_str(payload) {
                    Ok(v) => v,
                    Err(_) => continue,
                };
                match v.get("type").and_then(|t| t.as_str()) {
                    Some("content_block_delta") => {
                        if let Some(text) = v
                            .get("delta")
                            .and_then(|d| d.get("text"))
                            .and_then(|t| t.as_str())
                        {
                            sink(AiEvent::Delta(text.to_string()));
                        }
                    }
                    Some("message_start") => {
                        if let Some(u) = v.get("message").and_then(|m| m.get("usage")) {
                            in_tokens = u
                                .get("input_tokens")
                                .and_then(|x| x.as_u64())
                                .unwrap_or(0) as u32;
                            out_tokens = u
                                .get("output_tokens")
                                .and_then(|x| x.as_u64())
                                .unwrap_or(0) as u32;
                        }
                    }
                    Some("message_delta") => {
                        if let Some(u) = v.get("usage") {
                            if let Some(o) =
                                u.get("output_tokens").and_then(|x| x.as_u64())
                            {
                                out_tokens = o as u32;
                            }
                        }
                    }
                    Some("message_stop") => break,
                    _ => {}
                }
            }
        }

        let cost = estimate_cost("anthropic", &req.model, in_tokens, out_tokens);
        sink(AiEvent::Done(Usage {
            in_tokens,
            out_tokens,
            cost_usd: cost,
        }));
        Ok(())
    }

    fn list_models(&self) -> Result<Vec<ModelInfo>> {
        #[derive(Deserialize)]
        struct ModelsResp {
            data: Vec<ModelEntry>,
        }
        #[derive(Deserialize)]
        struct ModelEntry {
            id: String,
            display_name: Option<String>,
        }

        let resp = self
            .client
            .get(format!("{BASE}/models"))
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", ANTHROPIC_VERSION)
            .send()
            .context("anthropic models")?;
        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().unwrap_or_default();
            return Err(anyhow!("anthropic models {}: {}", status, text));
        }
        let parsed: ModelsResp = resp.json().context("parse models")?;
        Ok(parsed
            .data
            .into_iter()
            .map(|m| ModelInfo {
                name: m.display_name.clone().unwrap_or_else(|| m.id.clone()),
                id: m.id,
            })
            .collect())
    }
}
